knora/dsplib/utils/xml_upload.py

```diff
@@ -910,94 +918,340 @@ def xml_upload(input_file: str, server: str, user: str, password: str, imgdir: s
         if resource.ark:
             resource_iri = convert_ark_v0_to_resource_iri(resource.ark)
 
+        # in case of a multimedia resource: upload the multimedia file
         resource_bitstream = None
         if resource.bitstream:
-            img = sipi_server.upload_bitstream(os.path.join(imgdir, resource.bitstream.value))
-            internal_file_name_bitstream = img['uploadedFiles'][0]['internalFilename']
-            resource_bitstream = resource.get_bitstream(internal_file_name_bitstream, permissions_lookup)
+            try:
+                img = try_sipi_upload(sipi_server=sipi_server, filepath=os.path.join(imgdir, resource.bitstream.value))
+                internal_file_name_bitstream = img['uploadedFiles'][0]['internalFilename']
+                resource_bitstream = resource.get_bitstream(internal_file_name_bitstream, permissions_lookup)
+            except BaseError:
+                print(f"ERROR while trying to create resource '{resource.label}' ({resource.id}). ")
+                failed_uploads.append(resource.id)
+                continue
 
         permissions_tmp = permissions_lookup.get(resource.permissions)
 
-        try:
-            # create a resource instance (ResourceInstance) from the given resource in the XML (XMLResource)
-            resclass_type = resclass_name_2_type[resource.restype]
-            properties = resource.get_propvals(res_iri_lookup, permissions_lookup)
-            resclass_instance: ResourceInstance = resclass_type(
-                con=con,
-                label=resource.label,
-                iri=resource_iri,
-                permissions=permissions_tmp,
-                bitstream=resource_bitstream,
-                values=properties
-            )
-            resclass_instance = resclass_instance.create()
-        except BaseError as err:
-            print(f"ERROR while trying to create resource '{resource.label}' ({resource.id}). "
-                  f"The error message was: {err.message}")
-            failed_uploads.append(resource.id)
-            continue
-        except Exception as exception:
-            print(f"EXCEPTION while trying to create resource '{resource.label}' ({resource.id}). "
-                  f"The exception message was: {exception}")
+        # create the resource in knora
+        resclass_instance: ResourceInstance = None
+        for _ in range(5):
```

Comment:
this whole bit is essentially the same as above, so this could probably be refactored to get rid of the duplication.

Reply:
I wonder how I could refactor this... I don't see a simple solution.

Reply:
in short: by splitting everything up into smaller methods: then you could have a `retry()` method that takes a function as a parameter, which it then calls repeatedly.

But yes, it's not trivial - and certainly not necessary for this PR.